tests/msc2716_test.go

```diff
@@ -255,19 +255,87 @@ func TestImportHistoricalMessages(t *testing.T) {
 			})
 		})
 
+		t.Run("Backfill still works after many batches are imported", func(t *testing.T) {
+			t.Parallel()
+
+			roomID := as.CreateRoom(t, createPublicRoomOpts)
+			alice.JoinRoom(t, roomID, nil)
+
+			// Create some normal messages in the timeline
+			eventIDsBefore := createMessagesInRoom(t, alice, roomID, 2, "eventIDsBefore")
+			eventIdBefore := eventIDsBefore[len(eventIDsBefore)-1]
+			timeAfterEventBefore := time.Now()
+
+			// We chose the magic number 11 because Synapse currently limits the
+			// backfill extremities to 5. 10 also seemed like a round number someone
+			// could pick for other homeserver implementations so I just did 10+1 to
+			// make sure it also worked in that case.
+			//numBatches := 11
+			numBatches := 2
+			numHistoricalMessagesPerBatch := 100
+			// wait X number of ms to ensure that the timestamp changes enough for
+			// each of the historical messages we try to import later
+			time.Sleep(time.Duration(numBatches*numHistoricalMessagesPerBatch) * timeBetweenMessages)
+
+			// eventIDsAfter
+			createMessagesInRoom(t, alice, roomID, 2, "eventIDsAfter")
+
+			// Import a long chain of batches connected to each other.
+			// We want to make sure Synapse doesn't blow up after we import
+			// many messages.
+			var expectedEventIDs []string
+			var baseInsertionEventID string
+			nextBatchID := ""
+			for i := 0; i < numBatches; i++ {
+				insertTime := timeAfterEventBefore.Add(timeBetweenMessages * time.Duration(numBatches-numHistoricalMessagesPerBatch*i))
+				batchSendRes := batchSendHistoricalMessages(
+					t,
+					as,
+					roomID,
+					eventIdBefore,
+					nextBatchID,
+					createJoinStateEventsForBatchSendRequest([]string{virtualUserID}, insertTime),
+					createMessageEventsForBatchSendRequest([]string{virtualUserID}, insertTime, numHistoricalMessagesPerBatch),
+					// Status
+					200,
+				)
+				batchSendResBody := client.ParseJSON(t, batchSendRes)
+				// Make sure we see all of the historical messages
+				expectedEventIDs = append(expectedEventIDs, client.GetJSONFieldStringArray(t, batchSendResBody, "event_ids")...)
+				nextBatchID = client.GetJSONFieldStr(t, batchSendResBody, "next_batch_id")
+
+				// Grab the base insertione event ID to reference later in the marker event
+				if i == 0 {
+					baseInsertionEventID = client.GetJSONFieldStr(t, batchSendResBody, "base_insertion_event_id")
+				}
+			}
+
+			// Make sure we see the events at the very start of the message history
+			expectedEventIDs = append(expectedEventIDs, eventIDsBefore...)
```

Comment:
So this is new -> old?

Reply:
The order doesn't matter because it's just a list of expected ID's, `expectedEventIDs`

We also just use it in check off function (`paginateUntilMessageCheckOff`) which doesn't care about the order.

---

There is a similar variable in another test where order does matter but we call it `expectedEventIDOrder`